internal/session/session.go

```diff
@@ -48,144 +71,151 @@ type Body struct {
 }
 
 func (s *Session) keepAlive() {
-	s.startWriter()
-	s.startReader()
+	go s.startWriter()
+	go s.startReader()
 }
 
 func (s *Session) startWriter() {
-	go func() {
-		for {
-			if s.serverClosed {
-				log.Logger.Debugf("session writer: session closed, closing writer...")
-				if s.writerClosedCh != nil {
-					s.writerClosedCh <- true
-					log.Logger.Debugf("session writer: writer chan closed.")
-				}
-				return
-			}
-			serverUrl := fmt.Sprintf("https://%s/api/v1/session", s.endpoint)
+	ticker := time.NewTicker(1)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-s.ctx.Done():
+			log.Logger.Debug("session writer: closing writer")
+			s.writerClosedCh <- true
+			log.Logger.Debug("session writer: closed writer")
+			return
+
+		case <-ticker.C:
+			ticker.Reset(s.pipeInterval)
+		}
 
-			reader, writer := io.Pipe()
-			req, err := http.NewRequest("POST", serverUrl, reader)
-			if err != nil {
-				log.Logger.Errorf("session writer: error creating request: %v, retrying in 3s...", err)
-				time.Sleep(3 * time.Second)
-				continue
-			}
-			req.Header.Set("machine_id", s.machineID)
-			req.Header.Set("session_type", "write")
-
-			log.Logger.Debugf("session writer: created")
-			goroutineCloseCh := make(chan any)
-			go func() {
-				for {
-					select {
-					case <-s.writerCloseCh:
-						writer.Close()
-					case <-goroutineCloseCh:
-						return
-					case body := <-s.writer:
-						bytes, err := json.Marshal(body)
-						if err != nil {
-							log.Logger.Errorf("session writer: failed to marshal body: %v", err)
-							continue
-						}
-						if _, err := writer.Write(bytes); err != nil {
-							log.Logger.Errorf("session writer: failed to write to pipe: %v", err)
-							continue
-						}
+		reader, writer := io.Pipe()
+		goroutineCloseCh := make(chan any)
+		go func() {
+			log.Logger.Debug("session writer created")
+			for {
+				select {
+				case <-s.writerCloseCh:
+					log.Logger.Debug("session writer closed")
+					writer.Close()
+					return
```

Comment:
@cardyok We didn't `return` before, now we do. Please check.

Reply:
similarly, here is forcing writer's goroutine to be closed after reader finishes closing, but i think returning here should be fine.